File: lib/scanLite.ts
const CHAIN_IDS: Record<string, string> = {
  ethereum: "1",
  eth: "1",
  bsc: "56",
  base: "8453",
  arbitrum: "42161",
  polygon: "137",
  optimism: "10",
};

export type ScanResult = {
  address: string;
  symbol?: string;
  verdict: string;
  summary: string;
  url?: string;
};

async function getJson(url: string): Promise<any> {
  try {
    const response = await fetch(url, {
      headers: { Accept: "application/json", "User-Agent": "flopdesk/1.0" },
      cache: "no-store",
      signal: AbortSignal.timeout(12000),
    });
    if (!response.ok) return null;
    return await response.json();
  } catch {
    return null;
  }
}

function flag(info: Record<string, unknown>, key: string): string {
  const value = String(info[key] ?? "").trim().toLowerCase();
  if (["1", "true", "yes"].includes(value)) return "yes";
  if (["0", "false", "no"].includes(value)) return "no";
  return value || "n/a";
}

async function bestPair(address: string): Promise<any | null> {
  const wanted = address.toLowerCase();
  let payload = await getJson(`https://api.dexscreener.com/latest/dex/tokens/${address}`);
  let pairs = Array.isArray(payload?.pairs) ? payload.pairs : [];
  if (!pairs.length) {
    payload = await getJson(`https://api.dexscreener.com/latest/dex/search?q=${address}`);
    pairs = Array.isArray(payload?.pairs) ? payload.pairs : [];
  }
  const ranked: Array<[number, number, any]> = [];
  for (const pair of pairs) {
    const base = String(pair?.baseToken?.address || "").toLowerCase();
    const quote = String(pair?.quoteToken?.address || "").toLowerCase();
    const role = base === wanted ? 2 : quote === wanted ? 1 : 0;
    if (!role) continue;
    ranked.push([role, Number(pair?.liquidity?.usd || 0), pair]);
  }
  ranked.sort((a, b) => b[0] - a[0] || b[1] - a[1]);
  return ranked[0]?.[2] || null;
}

export async function scanToken(address: string): Promise<ScanResult> {
  let pair = await bestPair(address);
  if (!pair) {
    await new Promise((resolve) => setTimeout(resolve, 800));
    pair = await bestPair(address);
  }
  if (!pair) {
    return {
      address,
      verdict: "unknown",
      summary: `SCAN ${address} | no DEX pair found | verdict=unknown`,
    };
  }
  const wanted = address.toLowerCase();
  const isBase = String(pair.baseToken?.address || "").toLowerCase() === wanted;
  const token = isBase ? pair.baseToken : pair.quoteToken;
  let price = pair.priceUsd || "n/a";
  if (!isBase) {
    try {
      price = (Number(pair.priceUsd) / Number(pair.priceNative)).toPrecision(6);
    } catch {
      price = "n/a";
    }
  }
  const chain = String(pair.chainId || "");
  const liq = Number(pair.liquidity?.usd || 0);
  const chainId = CHAIN_IDS[chain.toLowerCase()];
  let security: Record<string, unknown> = {};
  if (chainId) {
    const gp = await getJson(
      `https://api.gopluslabs.io/api/v1/token_security/${chainId}?contract_addresses=${address}`,
    );
    const info = gp?.result?.[wanted];
    if (info && typeof info === "object") security = info as Record<string, unknown>;
  }
  const honeypot = flag(security, "is_honeypot");
  const cannotSell = flag(security, "cannot_sell_all");
  const buyTax = security.buy_tax ?? "n/a";
  const sellTax = security.sell_tax ?? "n/a";
  const buy = Number(buyTax);
  const sell = Number(sellTax);
  let verdict = "ok";
  if (honeypot === "yes" || cannotSell === "yes") verdict = "danger";
  else if (buy >= 10 || sell >= 10 || liq < 5000) verdict = "caution";
  return {
    address,
    symbol: token?.symbol,
    verdict,
    url: pair.url,
    summary: `SCAN ${address} | ${token?.symbol || "?"} ${chain} | $${price} liq $${liq.toLocaleString("en-US", { maximumFractionDigits: 0 })} | honeypot=${honeypot} tax=${buyTax}/${sellTax} | verdict=${verdict} | ${pair.url || ""}`.slice(0, 4096),
  };
}

File: agent/scan_lite.py
"""Fast public token check: Dexscreener + GoPlus. No private keys."""

from __future__ import annotations

import json
import re
from typing import Any
from urllib.parse import urlencode
from urllib.request import Request, urlopen

CHAIN_IDS = {
    "ethereum": "1",
    "eth": "1",
    "bsc": "56",
    "base": "8453",
    "arbitrum": "42161",
    "polygon": "137",
    "optimism": "10",
}

EVM_RE = re.compile(r"0x[a-fA-F0-9]{40}")


def _get_json(url: str, timeout: float = 15.0) -> Any:
    request = Request(url, headers={"Accept": "application/json", "User-Agent": "flopdesk/1.0"})
    try:
        with urlopen(request, timeout=timeout) as response:
            return json.loads(response.read().decode("utf-8"))
    except Exception:
        return None


def extract_contract(text: str) -> str | None:
    match = EVM_RE.search(text or "")
    if match:
        return match.group(0)
    return None


def _best_pair(address: str) -> dict[str, Any] | None:
    payload = _get_json(f"https://api.dexscreener.com/latest/dex/tokens/{address}") or {}
    pairs = payload.get("pairs") if isinstance(payload, dict) else None
    if not isinstance(pairs, list) or not pairs:
        search = _get_json(f"https://api.dexscreener.com/latest/dex/search?q={address}") or {}
        pairs = search.get("pairs") if isinstance(search, dict) else None
    if not isinstance(pairs, list) or not pairs:
        return None
    wanted = address.lower()
    ranked: list[tuple[int, float, dict[str, Any]]] = []
    for pair in pairs:
        if not isinstance(pair, dict):
            continue
        base = str((pair.get("baseToken") or {}).get("address") or "").lower()
        quote = str((pair.get("quoteToken") or {}).get("address") or "").lower()
        if base == wanted:
            role = 2
        elif quote == wanted:
            role = 1
        else:
            continue
        liquidity = ((pair.get("liquidity") or {}).get("usd")) or 0
        try:
            liq = float(liquidity)
        except (TypeError, ValueError):
            liq = 0.0
        ranked.append((role, liq, pair))
    if not ranked:
        return None
    ranked.sort(key=lambda item: (item[0], item[1]), reverse=True)
    return ranked[0][2]


def _goplus(chain: str, address: str) -> dict[str, Any]:
    chain_id = CHAIN_IDS.get(chain.lower())
    if not chain_id:
        return {}
    query = urlencode({"contract_addresses": address})
    payload = _get_json(f"https://api.gopluslabs.io/api/v1/token_security/{chain_id}?{query}")
    result = payload.get("result") if isinstance(payload, dict) else None
    if isinstance(result, dict):
        info = result.get(address.lower())
        if isinstance(info, dict):
            return info
    return {}


def _flag(info: dict[str, Any], key: str) -> str:
    value = str(info.get(key) or "").strip().lower()
    if value in {"1", "true", "yes"}:
        return "yes"
    if value in {"0", "false", "no"}:
        return "no"
    return value or "n/a"


def scan_token(address: str) -> dict[str, Any]:
    pair = _best_pair(address)
    if pair is None:
        return {
            "address": address,
            "verdict": "unknown",
            "summary": f"SCAN {address} | no DEX pair found | verdict=unknown",
        }
    base = pair.get("baseToken") or {}
    quote = pair.get("quoteToken") or {}
    if str(base.get("address") or "").lower() == address.lower():
        token = base
        price = pair.get("priceUsd") or "n/a"
    else:
        token = quote
        try:
            price = f"{float(pair.get('priceUsd')) / float(pair.get('priceNative')):.6g}"
        except (TypeError, ValueError, ZeroDivisionError):
            price = "n/a"
    chain = str(pair.get("chainId") or "")
    liquidity = ((pair.get("liquidity") or {}).get("usd")) or 0
    security = _goplus(chain, address)
    honeypot = _flag(security, "is_honeypot")
    cannot_sell = _flag(security, "cannot_sell_all")
    buy_tax = security.get("buy_tax") or "n/a"
    sell_tax = security.get("sell_tax") or "n/a"
    try:
        liq = float(liquidity)
    except (TypeError, ValueError):
        liq = 0.0
    try:
        buy = float(buy_tax)
        sell = float(sell_tax)
    except (TypeError, ValueError):
        buy, sell = 0.0, 0.0
    verdict = "ok"
    if honeypot == "yes" or cannot_sell == "yes":
        verdict = "danger"
    elif buy >= 10 or sell >= 10 or liq < 5000:
        verdict = "caution"
    summary = (
        f"SCAN {address} | {token.get('symbol') or '?'} {chain} | "
        f"${price} liq ${liq:,.0f} | honeypot={honeypot} tax={buy_tax}/{sell_tax} | "
        f"verdict={verdict} | {pair.get('url') or ''}"
    )
    return {
        "address": address,
        "symbol": token.get("symbol"),
        "name": token.get("name"),
        "chain": chain,
        "price": price,
        "liquidity": liq,
        "honeypot": honeypot,
        "buy_tax": buy_tax,
        "sell_tax": sell_tax,
        "verdict": verdict,
        "url": pair.get("url"),
        "summary": summary[:4096],
    }
